lib/puppet/util/json.rb

```diff
@@ -0,0 +1,57 @@
+module Puppet::Util
+  module Json
+    class ParseError < StandardError
+    end
+
+    begin
+      require 'multi_json'
+      # Force backend detection before attempting to use the library
+      # or load any other JSON libraries
+      MultiJson.default_adapter
+
+      # Preserve core type monkey-patching done by the built-in JSON gem
+      require 'json'
+    rescue LoadError
+      require 'json'
+    end
+
+    # These methods do similar processing to the fallback implemented by MultiJson
+    # when using the built-in JSON backend, to ensure consistent behavior
+    # whether or not Puppet::Util::Json can be loaded.
+    def self.load(string, options = {})
+      if defined? MultiJson
+        begin
+          MultiJson.load(string, options)
+        rescue MultiJson::ParseError => e
+          raise Puppet::Util::Json::ParseError, e.cause
+        end
+      else
+        begin
+          if string.nil?
+            raise Puppet::Util::Json::ParseError, "Invalid JSON: nil input"
+          end
+
+          string = string.read if string.respond_to?(:read)
+
+          if string.respond_to?(:force_encoding)
+            string = string.dup.force_encoding(::Encoding::ASCII_8BIT)
```

Comment:
Why would we do this?

Reply:
I wondered about this too, put it in because I didn't want surprises when we actually used MultiJson, which does it here https://github.com/intridea/multi_json/commit/8695900bd29d15b019dc82b0bd04830d78f2d0b0. Unfortunately there's no explanation as to why...

Reply:
That's wild. I'm still a bit nervous because of https://github.com/sensu/sensu/issues/783 and think it would be a good idea to find out _why_ this line is important. Though from irb it looks like this does what we want:
```rb
irb(main):001:0> s = File.read 'a'
=> "{\n  \"author\": \"ර\"\n}\n"

irb(main):003:0> s.encoding
=> #<Encoding:UTF-8>

irb(main):005:0> JSON.parse s
=> {"author"=>"ර"}

irb(main):016:0* s.force_encoding(::Encoding::ASCII_8BIT)
=> "{\n  \"author\": \"\xE0\xB6\xBB\"\n}\n"

irb(main):017:0> _.encoding
=> #<Encoding:ASCII-8BIT>

irb(main):008:0> JSON.parse s.force_encoding(::Encoding::ASCII_8BIT)
=> {"author"=>"ර"}

irb(main):009:0> _['author']
=> "ර"

irb(main):010:0> _.encoding
=> #<Encoding:UTF-8>
```

Reply:
Note that the [yajl adapter](https://github.com/intridea/multi_json/blob/master/lib/multi_json/adapters/yajl.rb) doesn't do any force encoding, and that's the one they were using in that issue. Maybe this _fixes_ that problem but never made it to the yajl adapter.

Reply:
I don't understand this. This is on the branch of the if where we aren't even using multijson? When can that happen? The comment suggests that sometimes we'd be unable to load this file but still want to continue. When and why? If it's reasonable, then why do we introduce new encoding behavior we didn't have before when using the regular json gem?

Reply:
We won't use MultiJson when someone installs a newer agent with this code running alongside an pre-5.5.0 puppetserver. That puppetserver won't have the multijson gem (puppetserver 5.3.0 and above will be able to rely on the agent to supply it, see https://github.com/puppetlabs/puppetserver/pull/1647).

I could go either way on the encoding thing. I was just trying to be consistent with what MultiJson will do when we use it.